sway/desktop/render.c

```diff
@@ -0,0 +1,895 @@
+#include "sway/output.h"
+#include "sway/server.h"
+#include "sway/tree/container.h"
+#include "sway/tree/layout.h"
+#include "sway/tree/view.h"
+#include "sway/tree/workspace.h"
+
+struct render_context {
+	struct sway_output *output;
+	double output_lx;
+	double output_ly;
+	pixman_region32_t *damage;
+	float matrix[9];
+	float scale;
+};
+struct render_context context;
+
+static void scale_box(struct wlr_box *box, float scale) {
+	box->x *= scale;
+	box->y *= scale;
+	box->width *= scale;
+	box->height *= scale;
+}
+
+static void scissor_output(struct wlr_output *wlr_output,
+		pixman_box32_t *rect) {
+	struct wlr_renderer *renderer = wlr_backend_get_renderer(wlr_output->backend);
+	assert(renderer);
+
+	struct wlr_box box = {
+		.x = rect->x1,
+		.y = rect->y1,
+		.width = rect->x2 - rect->x1,
+		.height = rect->y2 - rect->y1,
+	};
+
+	int ow, oh;
+	wlr_output_transformed_resolution(wlr_output, &ow, &oh);
+
+	enum wl_output_transform transform =
+		wlr_output_transform_invert(wlr_output->transform);
+	wlr_box_transform(&box, transform, ow, oh, &box);
+
+	wlr_renderer_scissor(renderer, &box);
+}
+
+static void render_texture(struct wlr_texture *texture,
+		const struct wlr_box *box, const float matrix[static 9], float alpha) {
+	struct wlr_output *wlr_output = context.output->wlr_output;
+	struct wlr_renderer *renderer =
+		wlr_backend_get_renderer(wlr_output->backend);
+
+	pixman_region32_t damage;
+	pixman_region32_init(&damage);
+	pixman_region32_union_rect(&damage, &damage, box->x, box->y,
+		box->width, box->height);
+	pixman_region32_intersect(&damage, &damage, context.damage);
+	bool damaged = pixman_region32_not_empty(&damage);
+	if (!damaged) {
+		goto damage_finish;
+	}
+
+	int nrects;
+	pixman_box32_t *rects = pixman_region32_rectangles(&damage, &nrects);
+	for (int i = 0; i < nrects; ++i) {
+		scissor_output(wlr_output, &rects[i]);
+		wlr_render_texture_with_matrix(renderer, texture, matrix, alpha);
+	}
+
+damage_finish:
+	pixman_region32_fini(&damage);
+}
+
+static void render_surface(struct wlr_surface *surface, int ox, int oy,
+		float alpha) {
+	struct wlr_texture *texture = wlr_surface_get_texture(surface);
+	if (!texture) {
+		return;
+	}
+
+	struct wlr_box surface_box = {
+		.x = ox,
+		.y = oy,
+		.width = surface->current->width,
+		.height = surface->current->height,
+	};
+	struct wlr_box output_box = {
+		.width = context.output->swayc->current.swayc_width,
+		.height = context.output->swayc->current.swayc_height,
+	};
+	struct wlr_box intersection;
+	if (!wlr_box_intersection(&output_box, &surface_box, &intersection)) {
+		return;
+	}
+
+	scale_box(&surface_box, context.scale);
+
+	float matrix[9];
+	enum wl_output_transform transform =
+		wlr_output_transform_invert(surface->current->transform);
+	wlr_matrix_project_box(matrix, &surface_box, transform, 0,
+		context.matrix);
+
+	render_texture(texture, &surface_box, matrix, alpha);
+}
+
+static void render_surfaces(struct wlr_surface *surface, double ox, double oy,
```

Comment:
This is just a `wlr_surface_for_each_surface` away. This function will get more complicated when we'll properly support subsurfaces in wlroots (which makes it possible to place subsurfaces behind their parent for instance).

Reply:
>I've mentioned this in #sway already, but I've found the iterator functions difficult to follow.

How iterators are currently used is over-complicated and could be simplified a lot.

>I know some developers who strictly believe in not having any duplicated code, and inadvertently add complexity to their code to avoid it.

I agree with you, but here I think iterators would be better.